### ISSUE TITLE
same pool for driving mercury and executing RPCs?

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-margo#5]
In GitLab by @jenkins on Feb 18, 2016, 12:35

Currently, margo is expected to run in its own ES - otherwise, other ULTS/tasks on the same ES can arbitrarily wait while the margo ULT calls progress.

Any possibility we can get the programming benefits of margo without having to tie ourselves to the thread-handoff model? I can imagine the following is a simple enough extension:

- margo enters "sharing is caring" mode when the progress pool and handler pool are the same
- only margo or ULTs/tasks created by margo may create ULTs/tasks in the shared pool
- when the margo instance is the only active ULT/task in the pool, margo is free to use blocking progression
- when the pool contains ULTs/tasks other than margo, margo uses non-blocking progress/trigger functions (i.e. uses a timeout of 0)

The downside is that we'll be spinning whenever we're processing operations. If only we had one event loop to rule them all...

**Comment by carns** [comment]
In GitLab by @carns on Feb 18, 2016, 13:43

This actually shouldn't be that hard, and it doesn't have to spin either.

I think the setup would look like this:
- from the caller's perspective, it is the same as usually except that you specify the same pool for both handlers and progress at init() time
- in the hg_progress_fn() we change this:

```
HG_Progress(mid->hg_class, mid->hg_context, 100);
```

Into this:

```
if(pool_thread_count > 1){
   HG_Progress(mid->hg_class, mid->hg_context, 0);
   ABT_thread_yield();
}
else {
   HG_Progress(mid->hg_class, mid->hg_context, 100);
}
```

The only minor unknown is how to find the thread count for the pool to know if we should yield or not.  There may be an ABT function for this, or maybe we can reference count explicitly somehow.

I'll have a look.

**Comment by carns** [comment]
In GitLab by @jenkins on Feb 18, 2016, 13:46

Looks like the count can be provided by ABT_pool_get_total_size (http://www.mcs.anl.gov/~sseo/public/argobots/d9/d52/group__POOL.html#ga89312324e2c5a199992323035d9bb7c2).

**Comment by carns** [comment]
In GitLab by @carns on Feb 18, 2016, 13:47

Sounds easy enough :)  I'll add a test program (or add it as an option to an existing test program) and try it out.

**Comment by carns** [comment]
In GitLab by @jenkins on Feb 18, 2016, 13:50

Awesome! As an aside, to test threading vs. non-threading in the leveldb wrappers, I effectively have two implementations, one using margo and one using callback functions. I would love to kill the latter :).

**Comment by carns** [comment]
In GitLab by @carns on Feb 18, 2016, 14:03

Status changed to closed by commit cab2fa11b8b119001aaeea446151b439cff58fdb
